### PR TITLE
Reduce shared memory usage storing the empty array

### PR DESCRIPTION
**Pull request by TysonAndre**
Reduce shared memory usage in php 7.3+ for `apc.serializer=default`. Store the pointer to the immutable zend_empty_array rather than a brand new array. (APCu is already storing a pointer to static variables elsewhere, e.g. uninitialized_bucket)

### CODE REVIEW

**Comment by nikic** [comment]
I don't think this is correct in this form. Not every empty array is a zend_empty_array.

Another thing to keep in mind is that empty arrays have have different next free elements, though I'd have to check whether apcu preserves that, possibly it's already ignored.

**Comment by TysonAndre** [comment]
I fixed the error - the earlier commit made this use 0 memory, but didn't set the zval to zend_empty_array when persisting.


> I don't think this is correct in this form. Not every empty array is a zend_empty_array.

I know, but the question is whether any empty array (in userland arrays) can be replaced with the zend_empty_array without observable effects. On unpersist, this is already using the EMPTY_ARRAY from #323 

```c
		case IS_ARRAY:
#if PHP_VERSION_ID >= 70300
			if (Z_ARR_P(zv)->nNumOfElements == 0) {
				ZVAL_EMPTY_ARRAY(zv); /* #323 */
				return;
			}
#endif
			Z_ARR_P(zv) = apc_unpersist_ht(ctxt, Z_ARR_P(zv));
			return;
```

> Another thing to keep in mind is that empty arrays have have different next free elements, though I'd have to check whether apcu preserves that, possibly it's already ignored.

apc_persist_copy_ht ignores that. serialize() functions also ignore that by design.

```c
	if (ht->nNumUsed == 0) {
#if PHP_VERSION_ID >= 70400
		ht->u.flags = HASH_FLAG_UNINITIALIZED;
#else
		ht->u.flags &= ~(HASH_FLAG_INITIALIZED|HASH_FLAG_PACKED);
#endif
		ht->nNextFreeElement = 0;
		ht->nTableMask = HT_MIN_MASK;
		HT_SET_DATA_ADDR(ht, &uninitialized_bucket);
		return ht;
	}
```

File: apc_persist.c [hunk 1]
```diff
@@ -127,10 +127,16 @@ static zend_bool apc_persist_calc_memoize(apc_persist_context_t *ctxt, void *ptr
 static zend_bool apc_persist_calc_ht(apc_persist_context_t *ctxt, const HashTable *ht) {
 	uint32_t idx;
 
+#if PHP_VERSION_ID < 70300
 	ADD_SIZE(sizeof(HashTable));
-	if (ht->nNumUsed == 0) {
+#endif
+	if (ht->nNumOfElements == 0) {
```
Comment (nikic) [inline apc_persist.c:133]: This is written in a confusing way. I'd move the condition ADD_SIZE in here...

File: apc_persist.c [hunk 2]
```diff
@@ -127,10 +127,16 @@ static zend_bool apc_persist_calc_memoize(apc_persist_context_t *ctxt, void *ptr
...
 		return 1;
 	}
+#if PHP_VERSION_ID >= 70300
+	/* In php 7.3+, this points to the immutable zend_empty_array outside of shared memory. */
+	ADD_SIZE(sizeof(HashTable));
```
Comment (nikic) [inline apc_persist.c:138]: ...and make this one unconditional.

**Pull request merged**